### DISCUSSION TITLE
type vs. interface

### DISCUSSION THREAD

**Discussion by uraflower** [KakaoFunding/front-end#66] [Polls]
`type`과 `interface` 중 어떤 것을 어떤 때에 사용할지 결정하면 좋겠습니다.
저는 두 개를 혼용하는 것보다 하나만 선택해서 통일성을 챙기고, 헷갈리지 않았으면 합니다.

## `type`과 `interface`의 차이점

- interface는 타입을 항상 객체로 사용해야 하며, type은 타입과 관계 없이 자유롭게 사용 가능하다(type alias).
- interface는 이름이 동일한 interface끼리 merge되는 특성이 있다. 반면 type은 이름이 동일한 type이 여러 개인 경우 에러가 발생한다.
- vscode에서 type으로 선언한 타입에 마우스를 호버하면 해당 타입의 내부 요소들을 확인할 수 있다. 그러나 interface로 선언한 타입의 내부 요소는 보여주지 않는다.
  
  - ![image](https://github.com/KakaoFunding/front-end/assets/82873315/1a18c037-059d-46ce-bf2e-f696e180d9bc)
  - ![image](https://github.com/KakaoFunding/front-end/assets/82873315/9a703af7-ac67-4b24-9e31-437974353251) 

- vscode에서 지원하는 타입스크립트 화살표 함수형 컴포넌트 스니펫(`tsrafce`)은 `props` 타입 지정 시 `type`을 사용한다. 개발 생산성과 연관되는 부분이다.

## 마무리

이게 **`type`과 `interface`의 모든 차이점은 아니라는 점** 알아주세요.
제가 찾아본 `type`과 `interface`의 차이점 중에서 중요하다고 생각되는 것들을 요약한 것일 뿐입니다 !

일반적으로 다른 사람들은 어떻게 생각하고 사용하는지 알기 위해서 유튜브, 레딧을 찾아봤는데 의견이 분분했습니다.

"'type'script이지 'interface'script가 아니다"라는 말도 있었고(조크겠지만), "`type`은 타이핑할 게 적다"는 의견도 있었습니다.
"객체에는 `interface`, 한 줄로 타입 지정이 필요하면 `type`을 사용한다"는 의견이 공감을 가장 많이 받는 것 같았습니다([링크](https://www.reddit.com/r/reactjs/comments/14q6nyr/interface_vs_type_what_should_i_use_in/)).

저는 vscode 스니펫과 interface merge 특성 때문에 `type`쪽에 마음이 더 기우는 것 같습니다.
질문이나 다른 의견 있으시면 댓글 남겨주세요 !

**Comment by ohgus** [comment]
만약 하나를 정해서 쓰자고 한다면 저는 `type`이 더 좋을거 같습니다.

**Comment by devkyoung2** [comment]
저는 혼용해서 쓰는것이 좋을것같아요. 나중에 타입을 확장할수도 있을것같은데 그냥 `type` 은 확장이 어려우니까요...
기본적으로 `type` 을 사용하되, 확장이 필요할 수도 있는 부분(api 응답)에서는 `interface`를 사용하는건 어떨까요 ?

**Reply by uraflower** [reply]
`type`이 확장이 어려운 경우가 혹시 어떤 경우인지 설명해주실 수 있을까요? 저는 확장 면에서는 둘 다 비슷하다고 생각했거든요

**Reply by devkyoung2** [reply]
아 제가 잘못봤네요 `&` 키워드로 확장이 가능하군요...! 처음알았네요!
그래도 굳이 하나만 따라가는것보단 둘다 쓰는게 좋을 것 같아요
인터페이스 머지 기능이 도움될수도 있고 실제 회사에서는 어떤걸 사용할지 모르니까요
기본적으로 `type` 을 사용하되, 인터페이스 머지가 필요한 경우 `interface`를 사용하는것이 좋을 것 같아요

**Reply by uraflower** [reply]
저는 `interface`의 머지 특성이 타입을 예측 불가능하게 만들어서 혼동을 줄 우려가 있다고 생각합니다. 중복 선언했을 때 이를 에러나 워닝으로 알려주지 않으니까요.
또 확장 대신 중복 선언으로 머지할 타당한 이유가 있을지 잘 모르겠어요. 뭔가 머지를 이용할만한 상황이 딱히 떠오르지 않는 거 같아요.
제 생각에 실제 회사에서는 회사by회사, 부서by부서일 거 같아서, 저희가 합리적이라고 판단되는 거로 골라도 좋다고 생각합니다!

**Reply by ohgus** [reply]
type을 고정해서 사용하고 추후에 꼭 interface를 사용해야 하는 상황이 생긴다면 그 때 고민해도 괜찮지 않을까 생각합니다!

**Reply by devkyoung2** [reply]
저도 기본적으로는 `type`을 사용하고 나중에 객체 안의 정의할 프로퍼티의 수가 많아진다면 그 때 `interface`를 고안해도 좋을 것 같아요~
